psi4/driver/p4util/python_helpers.py

```diff
@@ -185,3 +189,84 @@ def pcm_helper(block):
+    """For PsiAPI mode, forms a basis specification function from *block*
+    and associates it with keyword *key* under handle *name*. Registers
+    the basis spec with Psi4 so that it can be applied again to future
+    molecules. For usage, see mints2, mints9, and cc54 test cases.
+
+    """
+    key = key.upper()
+    name = ('anonymous' + str(random.randint(0, 999))) if name == '' else name
+    cleanbas = basname(name).replace('-', '')  # further remove hyphens so can be function name
+    block = filter_comments(block)
+    command_lines = re.split('\n', block)
+
+    symbol_re = re.compile(r'^\s*assign\s+(?P<symbol>[A-Z]{1,3})\s+(?P<basis>[-+*\(\)\w]+)\s*$', re.IGNORECASE)
+    label_re = re.compile(r'^\s*assign\s+(?P<label>(?P<symbol>[A-Z]{1,3})(?:(_\w+)|(\d+))?)\s+(?P<basis>[-+*\(\)\w]+)\s*$', re.IGNORECASE)
+    all_re = re.compile(r'^\s*assign\s+(?P<basis>[-+*\(\)\w]+)\s*$', re.IGNORECASE)
+    basislabel = re.compile(r'\s*\[\s*([-*\(\)\w]+)\s*\]\s*')
+
+    def anon(mol, role):
+        basstrings = {}
+
+        # Start by looking for assign lines, and remove them
+        leftover_lines = []
+        assignments = False
+        for line in command_lines:
+            if symbol_re.match(line):
+                m = symbol_re.match(line)
+                mol.set_basis_by_symbol(m.group('symbol'), m.group('basis'), role=role)
+                assignments = True
+
+            elif label_re.match(line):
+                m = label_re.match(line)
+                mol.set_basis_by_label(m.group('label'), m.group('basis'), role=role)
+                assignments = True
+
+            elif all_re.match(line):
+                m = all_re.match(line)
+                mol.set_basis_all_atoms(m.group('basis'), role=role)
+                assignments = True
+
+            else:
+                # Ignore blank lines and accumulate remainder
+                if line and not line.isspace():
+                    leftover_lines.append(line.strip())
+
+        # Now look for regular basis set definitions
+        basblock = list(filter(None, basislabel.split('\n'.join(leftover_lines))))
+        if len(basblock) == 1:
+            if not assignments:
+                # case with no [basname] markers where whole block is contents of gbs file
+                mol.set_basis_all_atoms(name, role=role)
+                basstrings[basname(name)] = basblock[0]
+            else:
+                message = ("Conflicting basis set specification: assign lines present but shells have no [basname] label.""")
+                raise TestComparisonError(message)
+        else:
+            # case with specs separated by [basname] markers
+            for idx in range(0, len(basblock), 2):
+                basstrings[basname(basblock[idx])] = basblock[idx + 1]
+
+        return basstrings
+    anon.__name__ = 'basisspec_psi4_yo__' + cleanbas
+    qcdb.libmintsbasisset.basishorde[name.upper()] = anon
+    core.set_global_option(key, name)
```

Comment:
The docstring should mention that the function does this. Also, what do you think about adding a kwarg to control this? Defaulting to True would be fine, but personally, I would prefer if the default was to skip `core.set_global_option()` that way api mode has less "magic" in the background. 
The call to `core.set_global_option` is handy, but I think it's a good idea to avoid implicitly messing with global state *especially* in apimode. If I was to read the docstring for this new function and try it out with something like this:

```python
mymol = # setup molecule, 
# setting options
psi4.core.set_global_option("BASIS","sto-3g") # set my default
psi4.basis_helper("""
…
""", name="custom-bs1")
psi4.basis_helper("""
…
""", name="custom-bs2")
scf_energies = {"default_basis": 0.0, "custom-bs1":0.0, "custom-bs1": 0.0}
scf_energies["deafult_basis"] = psi4.energy('scf', mol=mymol)
psi4.core.set_global_option("BASIS", "custom-bs1")
scf_energies["custom-bs1"] = psi4.energy('scf', mol=mymol)
psi4.core.set_global_option("BASIS", "custom-bs2")
scf_energies["custom-bs2"] = psi4.energy('scf', mol=mymol)

# look at the energies, make table, whatever
```

I may be surprised to see that the entries for "custom-bs2" and "default_basis" are the same, or worse I might not notice if the "custom-bs2" is some very small modification to sto-3g,  and I was expecting the energies to be close.